lib/api/src/entities/function/mod.rs

```diff
@@ -147,6 +153,71 @@ impl Function {
         Self(BackendFunction::new_typed_with_env(store, env, func))
     }
 
+    /// Creates a new async host `Function` (dynamic) with the provided signature.
+    ///
+    /// The provided closure returns a future that resolves to the function results.
+    /// When invoked synchronously (via [`Function::call`]) the future will run to
+    /// completion immediately. When invoked through [`Function::call_async`], the
+    /// future may suspend and resume according to JSPI semantics.
+    pub fn new_async<FT, F, Fut>(store: &mut impl AsStoreMut, ty: FT, func: F) -> Self
+    where
+        FT: Into<FunctionType>,
+        F: Fn(&[Value]) -> Fut + 'static + Send + Sync,
+        Fut: Future<Output = Result<Vec<Value>, RuntimeError>> + 'static + Send,
+    {
+        Self(BackendFunction::new_async(store, ty, func))
+    }
+
+    /// Creates a new async host `Function` (dynamic) with the provided signature
+    /// and environment.
+    pub fn new_with_env_async<FT, F, Fut, T: Send + 'static>(
+        store: &mut impl AsStoreMut,
+        env: &FunctionEnv<T>,
+        ty: FT,
+        func: F,
+    ) -> Self
+    where
+        FT: Into<FunctionType>,
+        F: Fn(FunctionEnvMut<T>, &[Value]) -> Fut + 'static + Send + Sync,
+        Fut: Future<Output = Result<Vec<Value>, RuntimeError>> + 'static + Send,
+    {
+        Self(BackendFunction::new_with_env_async(store, env, ty, func))
+    }
+
+    /// Creates a new async host `Function` from a native typed function.
+    ///
+    /// The future can return either the raw result tuple or any type that implements
+    /// [`IntoResult`] for the result tuple (e.g. `Result<Rets, E>`).
+    pub fn new_typed_async<F, Fut, Args, Rets, RetsAsResult>(
+        store: &mut impl AsStoreMut,
+        func: F,
+    ) -> Self
+    where
+        F: Fn(Args) -> Fut + 'static + Send + Sync,
+        Fut: Future<Output = RetsAsResult> + 'static + Send,
+        Args: WasmTypeList,
+        Rets: WasmTypeList,
+        RetsAsResult: IntoResult<Rets>,
+    {
+        Self(BackendFunction::new_typed_async(store, func))
+    }
+
+    /// Creates a new async host `Function` with an environment from a typed function.
+    pub fn new_typed_with_env_async<T: Send + 'static, F, Fut, Args, Rets, RetsAsResult>(
+        store: &mut impl AsStoreMut,
+        env: &FunctionEnv<T>,
+        func: F,
+    ) -> Self
+    where
+        F: Fn(FunctionEnvMut<T>, Args) -> Fut + 'static + Send + Sync,
+        Fut: Future<Output = RetsAsResult> + 'static + Send,
+        Args: WasmTypeList,
+        Rets: WasmTypeList,
+        RetsAsResult: IntoResult<Rets>,
+    {
+        Self(BackendFunction::new_typed_with_env_async(store, env, func))
+    }
```

Comment:
Again, do we really need duplicate constructors for all async functions? (again, probably yes)